airlock/file_browser_api.py

```diff
@@ -216,6 +217,44 @@ def build_string(node, indent):
         return "\n".join(build_string(self, ""))
 
 
+def scantree(root: Path) -> tuple[list[UrlPath], set[UrlPath]]:
+    """Use os.scandir to quickly walk a file tree.
+
+    Basically, its faster because it effectively just opens every directory,
+    not every file. And its in C code.  But that gives us whether the entry is
+    a file or a directory, which is all we need.
+
+    We are only really interested in file paths - those include any parent
+    directories we need for the tree for free.  However, we do need to manually
+    track empty directories, or else they will be excluded.
+    """
+
+    paths = []
+    directories = set()
+
+    def scan(current: str) -> int:
+        children = 0
+
+        for entry in os.scandir(current):
+            children += 1
+            path = UrlPath(entry.path).relative_to(root)
+
+            if entry.is_dir():
+                dir_children = scan(entry.path)
+                if dir_children == 0:
+                    # add empty dir to pathlist, or else it will not be shown
+                    paths.append(path)
+                    directories.add(path)
```

Comment:
So `directories` here is ONLY empty directories?

Reply:
yes, as that's the only leaf node directory that is not a file.